src/Channels/DefaultWritableBufferExtensions.cs

```diff
@@ -48,5 +49,101 @@ public static void Write(this WritableBuffer buffer, Span<byte> source)
+            buffer.Ensure(len);
+            buffer.Memory.Write<T>(BitConverter.IsLittleEndian ? Reverse(value) : value);
+            buffer.CommitBytes(len);
+        }
+
+        /// <summary>
+        /// Reads a structure of type T out of a buffer of bytes.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteLittleEndian<[Primitive]T>(this WritableBuffer buffer, T value) where T : struct
+        {
+            int len = Unsafe.SizeOf<T>();
+            buffer.Ensure(len);
+            buffer.Memory.Write<T>(BitConverter.IsLittleEndian ? value : Reverse(value));
+            buffer.CommitBytes(len);
+        }
+
+        /// <summary>
+        /// Reverses a primitive value - performs an endianness swap
+        /// </summary>
+        internal static unsafe T Reverse<[Primitive]T>(T value) where T : struct
+        {
+            // note: relying on JIT goodness here!
+            if (typeof(T) == typeof(byte) || typeof(T) == typeof(sbyte))
+            {
+                return value;
+            }
+            else if (typeof(T) == typeof(ushort) || typeof(T) == typeof(short))
+            {
+                ushort val = 0;
+                Unsafe.Write(&val, value);
+                val = (ushort)((val >> 8) | (val << 8));
+                return Unsafe.Read<T>(&val);
+            }
+            else if (typeof(T) == typeof(uint) || typeof(T) == typeof(int)
+                || typeof(T) == typeof(float))
+            {
+                uint val = 0;
+                Unsafe.Write(&val, value);
+                val = (val << 24)
+                    | ((val & 0xFF00) << 8)
+                    | ((val & 0xFF0000) >> 8)
+                    | (val >> 24);
+                return Unsafe.Read<T>(&val);
+            }
+            else if (typeof(T) == typeof(ulong) || typeof(T) == typeof(long)
+                || typeof(T) == typeof(double))
+            {
+                ulong val = 0;
+                Unsafe.Write(&val, value);
+                val = (val << 56)
+                    | ((val & 0xFF00) << 40)
+                    | ((val & 0xFF0000) << 24)
+                    | ((val & 0xFF000000) << 8)
+                    | ((val & 0xFF00000000) >> 8)
+                    | ((val & 0xFF0000000000) >> 24)
+                    | ((val & 0xFF000000000000) >> 40)
+                    | (val >> 56);
+                return Unsafe.Read<T>(&val);
+            }
+            else
+            {
+                // default implementation
+                int len = Unsafe.SizeOf<T>();
+                var val = stackalloc byte[len];
+                Unsafe.Write(val, value);
+                int to = len >> 1, dest = len - 1;
+                for (int i = 0; i < to; i++)
+                {
+                    var tmp = val[i];
+                    val[i] = val[dest];
+                    val[dest--] = tmp;
+                }
+                return Unsafe.Read<T>(val);
```

Comment:
Is the intent of `[Primitive]` to restrict to actual primitives or to include types like `S` (arbitrary, but containing no GC references)?

Reversing all the bytes of say `struct S { int x, y; }` is not how I expect endianness to be swapped. Shouldn't  `x` and `y` each be swapped individually with their relative field order retained? Of course, that isn't really feasible to do magically. Perhaps this should throw in default case and force me to write out individual, truly primitive fields in a deliberate order? I think that would also prevent StructLayout.Auto from causing different ordering on different platforms.